precompiles/parachain-staking/src/lib.rs

```diff
@@ -60,103 +82,71 @@ where
-			[0x97, 0x1d, 0x44, 0xc8] => Self::nominator_bond_more(&input[SELECTOR_SIZE_BYTES..])?,
-			_ => {
-				log::trace!(
-					target: "staking-precompile",
-					"Failed to match function selector in staking wrapper precompile"
-				);
-				return Err(ExitError::Other(
-					"No staking wrapper method at selector given selector".into(),
-				));
-			}
-		};
-
-		let outer_call: Runtime::Call = inner_call.into();
-		let info = outer_call.get_dispatch_info();
-
-		// Make sure enough gas
-		if let Some(gas_limit) = target_gas {
-			let required_gas = Runtime::GasWeightMapping::weight_to_gas(info.weight);
-			if required_gas > gas_limit {
-				return Err(ExitError::OutOfGas);
-			}
-		}
-		log::trace!(target: "staking-precompile", "Made it past gas check");
-
-		// Dispatch that call
-		let origin = Runtime::AddressMapping::into_account_id(context.caller);
-
-		log::trace!(target: "staking-precompile", "Gonna call with origin {:?}", origin);
-
-		match outer_call.dispatch(Some(origin).into()) {
-			Ok(post_info) => {
-				let gas_used = Runtime::GasWeightMapping::weight_to_gas(
-					post_info.actual_weight.unwrap_or(info.weight),
-				);
-				Ok(PrecompileOutput {
-					exit_status: ExitSucceed::Stopped,
-					cost: gas_used,
-					output: Default::default(),
-					logs: Default::default(),
-				})
-			}
-			Err(e) => {
-				let error_message = format!(
-					"Parachain staking call via EVM failed with dispatch error: {:?}",
-					e
-				);
-				Err(error(error_message))
-			}
+		let mut input = EvmDataReader::new(input);
+
+		match &input.read_selector()? {
+			// constants
+			Action::MinNomination => Self::min_nomination(target_gas),
+			// storage getters
+			Action::Points => Self::points(input, target_gas),
+			// role verifiers
+			Action::IsNominator => Self::is_nominator(input, target_gas),
+			Action::IsCandidate => Self::is_candidate(input, target_gas),
+			Action::IsSelectedCandidate => Self::is_selected_candidate(input, target_gas),
+			// runtime methods (dispatchables)
+			Action::JoinCandidates => Self::join_candidates(input, target_gas, context),
+			Action::LeaveCandidates => Self::leave_candidates(input, target_gas, context),
+			Action::GoOffline => Self::go_offline(target_gas, context),
+			Action::GoOnline => Self::go_online(target_gas, context),
+			Action::CandidateBondLess => Self::candidate_bond_less(input, target_gas, context),
+			Action::CandidateBondMore => Self::candidate_bond_more(input, target_gas, context),
+			Action::Nominate => Self::nominate(input, target_gas, context),
+			Action::LeaveNominators => Self::leave_nominators(input, target_gas, context),
+			Action::RevokeNomination => Self::revoke_nomination(input, target_gas, context),
+			Action::NominatorBondLess => Self::nominator_bond_less(input, target_gas, context),
+			Action::NominatorBondMore => Self::nominator_bond_more(input, target_gas, context),
+			_ => Err(error(
```

Comment:
From now on it is the read_selector function that handles the case where no selector matches. You should not use the default pattern, but list all the variants of your enum to make sure you handle all the cases.